lib/ecto/query/builder/filter.ex

```diff
@@ -65,6 +65,7 @@ defmodule Ecto.Query.Builder.Filter do
   end
 
   def escape(kind, expr, vars, env) when is_list(expr) do
+    assert_no_nils!(expr, Ecto.Query.CompileError)
```

Comment:
We need to add this check inside the map_reduce below, otherwise, if it is not a keyword list, but a list, it will fail with an weird error and the last clause currently in map_reduce would never be triggered. The same for `runtime!` below. We would also avoid traversing the list one extra time. :)